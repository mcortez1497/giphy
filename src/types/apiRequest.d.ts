interface ApiRequest {
  message?: string;
  isError: boolean;
  isLoading: boolean;
  isSuccess: boolean;
}

export { ApiRequest };
