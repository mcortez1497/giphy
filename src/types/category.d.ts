interface Category {
  _id: string;
  name: string;
}

export { Category };
