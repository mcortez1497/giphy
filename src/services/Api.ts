import { Dispatch } from "redux";

import { apiRequestStart, apiRequestSuccess, apiRequestError } from "reducers";
import { isObject } from "util";

interface FetchOptions {
  method?: "GET" | "POST" | "PATCH" | "PUT" | "DELETE";
  headers?: Headers;
  body?: string;
}

// Simple wrapper for the fetch API
// Dispatches actions so our app can know the status of
// each API request, as well as guarantee a formatted error object
class Api {
  private defaultOptions: FetchOptions = {
    headers: new Headers({
      "Content-Type": "application/json"
    })
  };

  public fetch(
    url: string,
    actionType: string,
    dispatch: Dispatch,
    options?: FetchOptions
  ): Promise<any> {
    const fetchOptions: FetchOptions = { ...this.defaultOptions, ...options };

    dispatch(apiRequestStart(actionType));

    return fetch(url, fetchOptions)
      .then(async response => {
        if (response.ok) {
          dispatch(apiRequestSuccess(actionType));
          return response.json();
        } else {
          const error = await response.json();
          dispatch(apiRequestError(actionType, this.buildErrorMessage(error)));
          return error;
        }
      })
      .catch(error => {
        dispatch(apiRequestError(actionType, this.buildErrorMessage(error)));
        return error;
      });
  }

  private buildErrorMessage = (error: any) =>
    isObject(error) && error.message
      ? error.message
      : "An unexpected error occurred.";
}

const ApiSingleton = new Api();
export { ApiSingleton as Api };
