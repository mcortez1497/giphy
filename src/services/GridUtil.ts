import { Gif } from "types";

interface GifBuckets {
  [key: number]: Gif[];
}

class GridUtil {
  // Calculate the total height of the container so the masonry effect will
  // fit properly into the calculated columns.
  public calculateContainerHeight = (
    numOfColumns: number,
    gifs: Gif[],
    includeCardFooters: boolean
  ) => {
    const cardFooterHeight = includeCardFooters ? 64 : 0;
    const cardMarginHeight = 8;

    // Total height of each card calulated by:
    // Card footer (64px) + margin (8px) + Gif Height * 2
    // We double height since each Gif is also being doubled in width within the Card
    const height = gifs.reduce((height: number, gif: Gif) => {
      return (
        height +
        cardFooterHeight +
        cardMarginHeight +
        parseInt(gif.height, 10) * 2
      );
    }, 0);

    // Divide the total height by the number of columns
    // Then add a buffer to get container height
    return height / numOfColumns + height / (1 + gifs.length) + "px";
  };

  // Divide Gifs evenly into buckets, so that whenever they are rendered,
  // they will not move as more gifs are loaded. 
  // Each column of the grid corresponds to one of these buckets.
  public buildGifBuckets = (gifs: Gif[], numOfColumns: number) => {
    const buckets = this.buildEmptyBuckets(numOfColumns);

    return gifs.reduce((acc: GifBuckets, gif: Gif, index: number) => {
      const mod = index % numOfColumns;
      acc[mod].push(gif);
      return acc;
    }, buckets);
  };

  private buildEmptyBuckets = (numOfColumns: number): GifBuckets =>
    Array(numOfColumns)
      .fill(0)
      .reduce((buckets: GifBuckets, _, index: number) => {
        buckets[index] = [];
        return buckets;
      }, {});
}

const GridUtilSingleton = new GridUtil();
export { GridUtilSingleton as GridUtil };
