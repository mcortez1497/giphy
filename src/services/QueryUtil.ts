import { QueryParams } from "types";

class QueryUtil {
  public parseQueryString = (queryString: string): QueryParams =>
    decodeURI(queryString)
      .substr(1)
      .split("&")
      .reduce((obj: QueryParams, query) => {
        const [key, value] = query.split("=");
        obj[key] = value;
        return obj;
      }, {});
}

const QueryUtilSingleton = new QueryUtil();
export { QueryUtilSingleton as QueryUtil };
