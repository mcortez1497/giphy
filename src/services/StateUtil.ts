import { AppState, defaultApiRequest } from "reducers";

class StateUtil {
  // Returns the ApiRequest object for an action, or a default object
  // if that API request hasn't been made yet.
  public getApiRequest = (state: AppState, apiRequestAction: string) =>
    state.api[apiRequestAction] || defaultApiRequest;

  // Returns the list of current "fresh" gifs merged with
  // currently saved gifs if there is any overlap
  public getFreshGifsMergedWithSavedGifs = (state: AppState) =>
    state.gifs.items.map(
      gif =>
        state.user.gifs.find(userGif => userGif.giphy_id === gif.giphy_id) ||
        gif
    );

  // Returns a list of saved gifs for a given category
  public getSavedGifsByCategory = (state: AppState, categoryId: string) =>
    state.user.gifs.filter(
      gif =>
        gif.categories &&
        gif.categories.find(category => category._id === categoryId)
    );

  // Returns a list of categories sorted alphabetically by name
  public getSortedCategories = (state: AppState) =>
    state.user.categories.items.sort((a, b) => (a.name > b.name ? 1 : -1));
}

const StateUtilSingleton = new StateUtil();
export { StateUtilSingleton as StateUtil };
