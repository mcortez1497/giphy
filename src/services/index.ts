export * from "./Api";
export * from "./GridUtil";
export * from "./QueryUtil";
export * from "./StateUtil";