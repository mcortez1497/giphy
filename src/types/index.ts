export * from "./category";
export * from "./gif";
export * from "./util";