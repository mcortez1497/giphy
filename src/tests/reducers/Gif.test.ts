import configureStore, { MockStoreEnhanced } from "redux-mock-store";
import thunkMiddleware, { ThunkDispatch } from "redux-thunk";

import {
  AppState,
  addGifs,
  setGifs,
  setPagination,
  getGifs,
  getMoreGifs,
  GifAction,
  gifReducer
} from "reducers";
import { Gif, GifPagination } from "types";
import { Api } from "services";

const middlewares = [thunkMiddleware];
const mockStore = configureStore(middlewares);

const initialState = {
  items: [],
  pagination: {
    count: 0,
    offset: 0,
    total_count: 0
  },
  query: ""
};

const responseMock = {
  data: [
    {
      id: "giphy_123",
      title: "test",
      height: "100",
      images: {
        fixed_width: {
          webp: "fixed_url"
        },
        original: {
          url: "original_url"
        }
      }
    }
  ],
  pagination: {
    count: 10,
    offset: 0,
    total_count: 100
  }
};

describe("Gif State", () => {
  describe("Actions", () => {
    it("returns addGifs action", () => {
      const gif: Gif = {
        giphy_id: "giphy_123",
        fixed_url: "fixed_url",
        original_url: "original_url",
        title: "test",
        height: "100",
        categories: []
      };
      expect(addGifs([gif])).toMatchSnapshot();
    });

    it("returns setGifs action", () => {
      const gif: Gif = {
        giphy_id: "giphy_123",
        fixed_url: "fixed_url",
        original_url: "original_url",
        title: "test",
        height: "100",
        categories: []
      };
      expect(setGifs([gif], "test")).toMatchSnapshot();
    });

    it("returns setPagination action", () => {
      const pagination: GifPagination = {
        count: 10,
        offset: 0,
        total_count: 100
      };
      expect(setPagination(pagination)).toMatchSnapshot();
    });
  });

  describe("Reducer", () => {
    it("add a gif", () => {
      const action = {
        type: "ADD_GIFS",
        items: [{
          giphy_id: "giphy_123",
          fixed_url: "fixed_url",
          original_url: "original_url",
          title: "test",
          height: "100",
          categories: []
        }]
      };
      expect(gifReducer(initialState, action)).toMatchSnapshot();
    });

    it("sets the array of gifs", () => {
      const action = {
        type: "SET_GIFS",
        items: [{
          giphy_id: "giphy_123",
          fixed_url: "fixed_url",
          original_url: "original_url",
          title: "test",
          height: "100",
          categories: []
        }],
        query: "test"
      };
      expect(gifReducer(initialState, action)).toMatchSnapshot();
    });

    it("sets pagination", () => {
      const action = {
        type: "SET_PAGINATION",
        pagination: {
          count: 10,
          offset: 0,
          total_count: 100
        }
      };
      expect(gifReducer(initialState, action)).toMatchSnapshot();
    });
  });

  describe("Thunks", () => {
    let store: MockStoreEnhanced<
      {},
      ThunkDispatch<AppState, undefined, GifAction>
    > = mockStore({
      gifs: initialState
    });

    beforeEach(() => {
      store.clearActions();
    });

    it("calls getGifs", () => {
      store = mockStore({
        gifs: initialState
      });

      Api.fetch = jest
        .fn()
        .mockImplementation((url: string) =>
          Promise.resolve(responseMock)
        );

      return store.dispatch(getGifs()).then(() => {
        expect(Api.fetch).toHaveBeenCalled();
        expect(store.getActions()).toMatchSnapshot();
      });
    });

    it("calls getMoreGifs", () => {
      store = mockStore({
        gifs: initialState
      });

      const responseData = {
        data: [
          {
            id: "giphy_456",
            title: "test",
            height: "100",
            images: {
              fixed_width: {
                webp: "fixed_url"
              },
              original: {
                url: "original_url"
              }
            }
          }
        ],
        pagination: {
          count: 10,
          offset: 0,
          total_count: 100
        }
      };

      Api.fetch = jest
        .fn()
        .mockImplementation((url: string) => Promise.resolve(responseData));

      return store.dispatch(getMoreGifs()).then(() => {
        expect(Api.fetch).toHaveBeenCalled();
        expect(store.getActions()).toMatchSnapshot();
      });
    });
  });
});
