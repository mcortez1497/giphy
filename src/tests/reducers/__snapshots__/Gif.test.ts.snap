// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gif State Actions returns addGifs action 1`] = `
Object {
  "items": Array [
    Object {
      "categories": Array [],
      "fixed_url": "fixed_url",
      "giphy_id": "giphy_123",
      "height": "100",
      "original_url": "original_url",
      "title": "test",
    },
  ],
  "type": "ADD_GIFS",
}
`;

exports[`Gif State Actions returns setGifs action 1`] = `
Object {
  "items": Array [
    Object {
      "categories": Array [],
      "fixed_url": "fixed_url",
      "giphy_id": "giphy_123",
      "height": "100",
      "original_url": "original_url",
      "title": "test",
    },
  ],
  "query": "test",
  "type": "SET_GIFS",
}
`;

exports[`Gif State Actions returns setPagination action 1`] = `
Object {
  "pagination": Object {
    "count": 10,
    "offset": 0,
    "total_count": 100,
  },
  "type": "SET_PAGINATION",
}
`;

exports[`Gif State Reducer add a gif 1`] = `
Object {
  "items": Array [
    Object {
      "categories": Array [],
      "fixed_url": "fixed_url",
      "giphy_id": "giphy_123",
      "height": "100",
      "original_url": "original_url",
      "title": "test",
    },
  ],
  "pagination": Object {
    "count": 0,
    "offset": 0,
    "total_count": 0,
  },
  "query": "",
}
`;

exports[`Gif State Reducer sets pagination 1`] = `
Object {
  "items": Array [],
  "pagination": Object {
    "count": 10,
    "offset": 0,
    "total_count": 100,
  },
  "query": "",
}
`;

exports[`Gif State Reducer sets the array of gifs 1`] = `
Object {
  "items": Array [
    Object {
      "categories": Array [],
      "fixed_url": "fixed_url",
      "giphy_id": "giphy_123",
      "height": "100",
      "original_url": "original_url",
      "title": "test",
    },
  ],
  "pagination": Object {
    "count": 0,
    "offset": 0,
    "total_count": 0,
  },
  "query": "test",
}
`;

exports[`Gif State Thunks calls getGifs 1`] = `
Array [
  Object {
    "items": Array [],
    "query": "",
    "type": "SET_GIFS",
  },
  Object {
    "items": Array [
      Object {
        "fixed_url": "fixed_url",
        "giphy_id": "giphy_123",
        "height": undefined,
        "original_url": "original_url",
        "title": "test",
      },
    ],
    "query": "",
    "type": "SET_GIFS",
  },
  Object {
    "pagination": Object {
      "count": 10,
      "offset": 0,
      "total_count": 100,
    },
    "type": "SET_PAGINATION",
  },
]
`;

exports[`Gif State Thunks calls getMoreGifs 1`] = `
Array [
  Object {
    "items": Array [
      Object {
        "fixed_url": "fixed_url",
        "giphy_id": "giphy_456",
        "height": undefined,
        "original_url": "original_url",
        "title": "test",
      },
    ],
    "type": "ADD_GIFS",
  },
  Object {
    "pagination": Object {
      "count": 10,
      "offset": 0,
      "total_count": 100,
    },
    "type": "SET_PAGINATION",
  },
]
`;
